neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,182 @@
+
+Turns out that the NEAR protocol is already solving a version of this problem.  State from all accounts is stored together in the same `Trie` data structure.  So in theory, if one account could figure out under what key another account has stored some value, it could issue a `storage_write` for that key and make unauthorised modifications to it.  To prevent this type of attack, when accounts issue a `storage_write`, the protocol prepends the `AccountId` of the account that is currently executing to the key.  This creates a unique namespace for each account that is inaccessible to all other accounts.
+
+We propose a similar approach for sharded contracts as well.  Instead of using the `AccountId` namespace, when an account starts using a sharded contract code, the protocol can create a new `ShardedContractInstanceId` for it.  This id can then be prepended to all storage accesses and just like above, this creates an isolated namespace for that specific sharded contract that the account cannot access.
+
+If an account stops using a sharded contract code (e.g. by deploying another contract code on itself), then the associated `ShardedContractInstanceId` can be removed.
+
+### Access control
+
+Sticking with the FT contract example, there are two types of access control checks the contract has to perform.
+
+1. When a sender wants to initiate a transfer.
+2. And when a receiver called by the sender account to accept some funds.
+
+The first case is fairly straight forward to address.  Only the owner of the account should be allowed to initiate a transfer.  As such, when the initiate transfer function is called, the smart contract can perform a check to ensure that the caller of the function is the owner of the account.  If this is not the case, then the call can be rejected.
+
+The second case is trickier to address.  In order to ensure that no tokens are maliciously minted, when a receiver is executing the call to receive tokens, it needs a guarantee that the sender of the message is indeed another instance of the sharded contract and that it has indeed decremented its local balance appropriately.  The minimum amount of information that the receiver needs to check to get this guarantee is that the receiver needs to know that the sender of the cross contract call was an instance of the sharded contract.
+
+Currently, the receiver of a message can check the `AccountId` of the sender of the message.  In this case that is not sufficient as the receiver does not know if the sender is indeed using an instance of the sharded contract or not.  Therefore, we propose that when an instance of a sharded contract sends a cross contract receipt to another contract, on top of including the `AccountId` of the actual account, we also include the `AccountId` of the sharded contract code.  Then the receiver of the cross contract receipt can check this bit of information and convince itself that this is a legitimate token transfer.
```

Comment:
It is more subtle than what is described here:
* For writes, it should be coming from another instance of the same contract (or the sharded contract account itself)
* For reads, it should allow any other contracts to call directly

Reply:
The most generic approach would be to specify explicit access for each method. For each method the developer could specify that it can be called by some subset of (anyone, owner, other contract instance).